### ISSUE TITLE
Design and agree on a 1.0 FFI

### ISSUE THREAD

**Issue by bendk** [mozilla/uniffi-rs#2155]
Lately, we've been discussing the idea of UniFFI 1.0 and what it would entail.  As part of this effort, we should think about how we could improve on our current FFI layer.  There's basically 2 reasons for this:

 - Historically, [UniFFI has favored shipping features quickly over performance](https://github.com/mozilla/uniffi-rs/blob/690fd55964122392ab7925f8f16f1a505ffdc110/docs/adr/0002-serialize-complex-datatypes.md), with the idea that we can revisit these decisions and improve on them later.  This seems like a great time to revisit some of those decisions and improve upon them.
- We have a lot of hindsight around implementing bindings generators which we can use to simplify things.  This will make our current bindings generation code more maintainable and make it easier for new bindings to be written.

General framework for making these decisions:
  - This issue will be act as a meta issue for this work and track decisions around the overall direction.  Please add your thoughts in the comments and we can update the plan as we go.
  - The goal is get all stakeholders to come to a consensus on an FFI that's "good enough".  This includes UniFFI users, UniFFI devs, bindings authors, developers of the ecosystems that UniFFI is being introduced into, etc.
  - Issues for specific tasks will be labeled with `FFI-1.0`.
  - There's no rush to get this done and we should expect that some of these issues will require slow and steady discussion before we can come to consensus.
  - Changes should be justified based on improve performance, simplify binding generation code, or both.
  - If a change is claimed to improve performance, there should be a benchmark added that measures this.

**Comment by bendk** [comment]
As mentioned in the description, this issue is a working document.  The above represents my understanding of the issue after talking it over with a couple other UniFFI devs.  If you disagree with anything or feel like something's missing, please add a comment.

**Comment by skeet70** [comment]
I agree that performance is something that should be optimized more before 1.0. Once we have a uniffi-bindgen-java version of our library that already exists natively in Java, we'll be benchmarking ourselves and may come up with ideas for improvements. Right now we have Rust, Python, and Kotlin benchmarks for that library but don't have a good comparison point to find improvements there.

Naively it seems to me like 1.0 should have either one of proc-macros or UDL but not both. I think that will simplify usage and make it easier to come up with a cohesive set of documentation.

Even though I'm implementing Java bindings, that's primarily translating Kotlin bindings, so I don't grok the bindgen internals well enough to give good improvement feedback there. I do think there's a lot to be gained in good documentation of what templates/code types need to be implemented and handled for bindings to support each feature UniFFI supports, but that's not really a maintenance improvement.

**Comment by Sajjon** [comment]
Hey! Exciting with `1.0` 🚀

I will re-iterate up [some things I wrote here](https://github.com/mozilla/uniffi-rs/pull/2087#issuecomment-2128711332), I think bindings can be improved quite a bit - also with multi-crate setup in mind. Instead of generating **huge** binding files - one per UniFFI crate (in a possible multi-UniFFI-crate-workspace repo) - also with multiple copies various UniFFI FfiConverters,  which at least is the case for Swift, we could generate many small files. Where meant to be internal symbols have appropriate access modifiers (`internal` in Swift), so that those symbols are visible across those files.

For Swift specifically we ought to re-design with Swift 6 language mode and `strict-concurrency` checking set to `complete` - because UniFFI generated Swift binding code does not compile, as I mentioned in https://github.com/mozilla/uniffi-rs/issues/2046. This is not a huge issue right now, but in September when _Xcode 16_ goes from beta to prod, and developers wanna adopt Swift 6 language mode, this will be a growing problem.

I think it is excellent timing to redesign the Swift bindings part of 1.0 of UniFFI to be prepared for Swift 6 and `complete` strict concurrency checking. I have some ideas, but I think it would be better to first make the changes to how bindgen generates code, into many files rather than one huge, and for multiple crates, ensure each UniFFI FfiConverter is only generated once, with `internal` as access modifier.

**Comment by mhammond** [comment]
> but I think it would be better to first make the changes to how bindgen generates code, into many files rather than one huge, and for multiple crates, ensure each UniFFI FfiConverter is only generated once, with internal as access modifier.

That can be done today, [all the components are passed to the swift generator](https://github.com/mozilla/uniffi-rs/blob/f4ca11f2ebc665efa8f959eab9b04e7e856f3688/uniffi_bindgen/src/bindings/swift/mod.rs#L82).

All the rest sounds great too!

**Comment by skeet70** [comment]
@Sajjon for reference `uniffi-bindgen-java` also generates individual files already.

I found out today about JNR. I think @badboy had a branch already trying out JNA direct mapping for performance benefits. Looking at [benchmarks](https://github.com/zakgof/java-native-benchmark) it may be worth trying out JNR. This would only help Kotlin/Java performance, not a general library improvement.

**Comment by alexkirsz** [comment]
I've already written up some of what we'd like to have in [this comment](https://github.com/mozilla/uniffi-rs/issues/1896#issuecomment-2395094512) and [this issue](https://github.com/mozilla/uniffi-rs/issues/2255), but I'll expand a bit more here.

We maintain a very large monorepo that contains code for multiple apps. Each app requires a particular set of features, but many of these features are shared between two or more apps. As such, we've split each feature into its own crates:

1. the API crate, which only contains the UniFFI interfaces (as traits) and types specific to this feature;
2. the implementation crate, which contains the actual implementation code for our traits; and
3. shared type library crates, which contain shared types between multiple features (e.g. a Vec2 type, or a Path type).

In order to keep compilation times to a minimum and avoid scaling issues, we run UniFFI bindgen on each library that exposes a UniFFI interface independently. This means that API crates, implementation crates, and type library crates will each invoke the UniFFI bindgen binary once, instead of a single invocation on the output shared library.

This results in the following compilation graph:

```
┌───────┐  ┌───────┐  ┌───────┐
│TYPES 1│  │TYPES 2│  │TYPES 3│
└───────┘  └───────┘  └───────┘
    │          │          │    
    └────┬─────┘          │    
         ▼                │    
      ┌─────┐             │    
      │API A│             │    
      └─────┘             │    
         │                │    
         ├────────────────┤    
         ▼                ▼    
     ┌──────┐          ┌─────┐ 
     │IMPL A│          │API B│ 
     └──────┘          └─────┘ 
         │                │    
         │                ▼    
         │           ┌────────┐
         └───────┐   │ IMPL B │
                 │   └────────┘
                 │        │    
                 ├────────┘    
                 ▼             
            ┌────────┐         
            │ TARGET │         
            └────────┘         
```

Each library has its own two-step compilation process:

1. Compile the library into a .rlib;
2. Codegen the library's foreign binding;
3. Compile the library's foreign binding.

And many such two-step compilation processes can be parallelized when there is no directed edge between their corresponding libraries.

This model also ensures that we can re-compile the foreign interface very fast, since API and types crates are guaranteed to only depend on each other, and not to include extraneous code that would otherwise increase the compilation type. As such, in most cases, iOS, Android, and Web engineers do not have to pay the cost of the implementation compilation times when working on features.

However, in order to make this work, we've made a small patch to UniFFI bindgen to allow it to read from multiple rlibs directly. For instance, in order to generate the foreign bindings for API A, we need to include the rlibs of TYPES 1, TYPES 2, and API A. This will properly generate the foreign code for API A, but will also re-generate the foreign bindings of TYPES 1 and TYPES 2, which we'll already have generated and compiled upstream.

Another issue is that all generated foreign bindings expect to be shipped as part of the same foreign library. For instance, in Swift, this means a single Swift library. As such, they all expect the other modules' symbols to already be imported in the current scope. However, in our case, these symbols need to be imported from the other Swift libraries: the foreign bindings for API A will need to `import Types1FFI` and `import Types2FFI`.

So in order to make this work, we:

1. Run UniFFI bindgen on all rlibs necessary for a given library; then
2. Select the generated binding file for the library, discarding the rest; then
3. Prepend necessary imports to the generated file.

The drawbacks of this approach are:

1. Some binding files will be generated many times, but only compiled once. For instance, in the graph above, the bindings of the TYPES 1 library will be generated 6 times: as part of its own compilation step, as well as part of any dependent library and all its transitive dependent libraries. However, it will only actually be compiled once, as part of its own compilation step. All the other generations are superfluous. This isn't too big of a problem for us at the moment since codegen is generally very fast, but as we scale to more features and more complex graphs, it might become one.
→ The bindgen binary should accept a parameter to select which module we want to generate bindings for.

2. Similarly, the rlibs need to be parsed multiple times by each UniFFI bindgen invocation.
→ The bindgen binary could generate an intermediate representation for UDLs and proc macro metadata. This would avoid repeating some of the parsing work, as well as embed additional metadata like the library name (see point 4.)

3. Each binding file will include its own RustBuffer implementation, as well as converter code for all the primitive types it uses.
→ The bindgen binary could expose a command to generate a library that contains all the boilerplate code, without any filtering. All generated binding files could then import this shared library.

4. Imports must be prepended as an additional step after the codegen.
→ The intermediate representation mentioned in 2. could include the expected import name/namespace of the dependency library, which would allow the codegen to both add the import (in Swift) and/or properly namespace any reference to a type coming from another library to avoid conflicts.

Currently, our implementation is tailored to our needs and, as such, cuts many corners and breaks some features we don't use. We would love to see UniFFI support this natively.

**Comment by bobozaur** [comment]
> However, in order to make this work, we've made a small patch to UniFFI bindgen to allow it to read from multiple rlibs directly. For instance, in order to generate the foreign bindings for API A, we need to include the rlibs of TYPES 1, TYPES 2, and API A. This will properly generate the foreign code for API A, but will also re-generate the foreign bindings of TYPES 1 and TYPES 2, which we'll already have generated and compiled upstream.

Just wanted to add here that we were able to make use of multiple rlibs by creating our own bindgen binary using lower level API from vanilla uniffi. Pretty sure that it might be possible to cache/import bindings generated upstream too, but, as @alexkirsz pointed out, code gen is fast enough for this not to be a concern. At least for now.

**Comment by bendk** [comment]
Hi all, thanks for all the thoughtful comments on this thread and sorry that I haven't been able to follow up on this yet.

It's interesting to me that a lot of these comments are about alternative ways to get the metadata.  That's not what comes to my mind when I think "FFI", but it makes perfect sense when we are loading the metadata from library files.  I think there's a lot to talk about here.  Should we split this into a separate issue?  One possibility that we've been toying with is parsing the metadata from the source directly, maybe using `rust-analyzer` (which handles a couple things better than `syn` in this scenario).

In the short-term, I'm hoping to do some serious work on the Kotlin FFI before the end of the year.  I just learned about JNR, maybe that could be useful.  Separately from improving performance, I'd really love to have an FFI system that's not tied to Java reflection.  It can lead to [weird issues](https://github.com/java-native-access/jna/issues/1686) that are hard to debug.

I also looked into JNI a bit and realized that if we went that road, we'd probably need to generate some additional Rust code, using the `jni` crate.  That's pretty similar to how we're looking into using `pyo3` to speed up the Python bindings.  Maybe we should dig deeper into this idea of bindings generating extra Rust code.

**Comment by skeet70** [comment]
For the java bindings we're probably going to use the newer Project Panama/FFM API when we do a performance pass. I've made two half-hearted attempts at it so far and it seems like it'll work. Haven't made it to the point of benchmarking vs existing java bindings or the Kotlin ones. We don't _need_ JVM 1.8/1.11 compatibility for our use case though, which will likely limit you to JNR being the best option.

**Comment by bendk** [comment]
I wonder if we could use libffi more-or-less directly with some jni shim code.  AFAICT, all the other solutions we're talking about depend on libffi and add an extra layer to the stack.  By removing that layer, we can speed things up, get more low-level control over what's happening, and maybe share the code between the Kotlin and Java bindings.

The main benefits of JNA/JNR/Panama is that they provide a higher-level interface, but that's kind of silly for the UniFFI use-case since we're also a code generator.  For example, it's cool to use reflection to define for FFI call and that's what I would use for hand-written code.  However, in our case, it wouldn't bee too hard to build up an array of arg types and call [ffi_prep_cif](https://www.chiark.greenend.org.uk/doc/libffi-dev/html/The-Basics.html).

Even more ambitious would be to ditch libffi entirely and have the JNI code generate everything needed for the bindings module.

**Comment by skeet70** [comment]
I believe Panama/FFM is as replacment for the JNI and in benchmarks I've seen frequently performs better. I don't think it's a JNI wrapper like the other tools.

-------- Original Message --------

**Comment by bendk** [comment]
[It seems like it uses JNI](https://github.com/search?q=repo%3Aopenjdk%2Fpanama-foreign%20jni&type=code).  Is there another low-level way for JVM code to make calls to a C library?

**Comment by skeet70** [comment]
As far as I'm aware, all those JNI mentions in the FFM repo are interoperability, glue, compatibility, etc. FFM _is_ the other low-level way [for JVM code to make calls to a C library](https://docs.oracle.com/en/java/javase/21/core/calling-c-library-function-foreign-function-and-memory-api.html#GUID-E7255CE9-5A95-437C-B37A-276B6C9B5F4D). Looking through the source of the core functionality of linking and calling C functions in the "normal" FFM way I didn't see any JNI references, but I certainly haven't done a super thorough dive to answer if it's used at all in any main-line patterns. At a glance it seems like some of the existing JNI code is used for error propagation or something for example, but for actually looking up and calling functions, managing memory, etc, new FFM code is used. Every JEP, blog post, example, etc I've interacted with about it has talked about it as a completely separate thing.

**Comment by paxbun** [comment]
The hits in the JNI search result in the `openjdk/panama-foreign` seem to be derived from `openjdk/jdk`. Searching `jni` in `openjdk/jdk` shows almost the same result. Though GitHub doesn't show that they're forked from `openjdk/jdk`, it appears that the experimental projects by OpenJDK fork the entire JDK repo.

AFAIK, both `Linker.downcallHandle` and `Linker.upcallStub` rely on internal JIT logic, and that's why `Linker.downcallHandle` returns `NativeMethodHandle extends MethodHandle` (it's an internal class) that holds a pointer to a JIT compiled code, and `Linker.upcallStub` returns `MemorySegment`, which is the representation of pointers on the Java side. I've also read some of the related code, and IIRC, they also used `MacroAssembler`, which is the C++ class that OpenJDK uses to implement JIT.

It seems FFM support in GraalVM has also graduated from the experimental state in GraalVM 25, so using FFM would be a better option for desktop JVM targets using the latest Java.

The problem is that Android is still using JDK 17 (and FFM is not supported), and there are still a few years left until the EOL of Java 8 and Java 17. Even if Android starts to support FFM sometime in the future, many Android Jetpack libraries will have minSdk = 21 (the latest is 36), and there will be many devices still not supporting FFM, so even if UniFFI chooses to use FFM, I hope JNI support will not be ceased as a backward compatibility feature.

**Comment by skeet70** [comment]
@bendk for reference I ported #2765 onto the Java bindings (not very surprisingly didn't notice performance changes since it's a stability change), made a couple not impactful specific-to-our-Java-bindings improvements (mostly pre-allocating things more), and then (proof of concept) had an AI convert from JNA to FFM/Panama.

I ran benches on our cryptography library that uses the bindings, as well as comparing them to benches of our older native Java library (TSC, which doesn't have a standalone mode for comparison unfortunately). All the non-standalone benches involve network calls to the same local service:

```
  FFM vs JNA
  ┌──────────────────────┬──────────┬──────────┬────────┐
  │      Operation       │ FFM (µs) │ JNA (µs) │ Change │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ TSP Encrypt          │          │          │        │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ 1B                   │ 549      │ 572      │ -4.0%  │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ 100B                 │ 541      │ 552      │ -1.9%  │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ 10KB                 │ 697      │ 646      │ +8.0%  │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ 1MB                  │ 8,743    │ 8,792    │ -0.6%  │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ TSP Decrypt          │          │          │        │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ 1B                   │ 543      │ 569      │ -4.6%  │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ 100B                 │ 533      │ 571      │ -6.6%  │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ 10KB                 │ 705      │ 640      │ +10.2% │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ 1MB                  │ 8,659    │ 9,261    │ -6.5%  │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ TSP Batch            │          │          │        │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ 10 docs × 100B       │ 1,204    │ 1,263    │ -4.7%  │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ Standalone Roundtrip │          │          │        │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ 10B                  │ 27       │ 97       │ -72.2% │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ 10KB                 │ 166      │ 248      │ -33.3% │
  ├──────────────────────┼──────────┼──────────┼────────┤
  │ 100KB                │ 1,351    │ 1,409    │ -4.1%  │
  └──────────────────────┴──────────┴──────────┴────────┘
  FFM vs Rust Native (FFI Overhead)
  ┌──────────────────────┬───────────┬──────────┬──────────┐
  │      Operation       │ Rust (µs) │ FFM (µs) │ Overhead │
  ├──────────────────────┼───────────┼──────────┼──────────┤
  │ Standalone Roundtrip │           │          │          │
  ├──────────────────────┼───────────┼──────────┼──────────┤
  │ 10B                  │ 7.6       │ 27       │ +254%    │
  ├──────────────────────┼───────────┼──────────┼──────────┤
  │ 10KB                 │ 99        │ 166      │ +68%     │
  ├──────────────────────┼───────────┼──────────┼──────────┤
  │ 100KB                │ 930       │ 1,351    │ +45%     │
  ├──────────────────────┼───────────┼──────────┼──────────┤
  │ TSP Encrypt          │           │          │          │
  ├──────────────────────┼───────────┼──────────┼──────────┤
  │ 1B                   │ 471       │ 549      │ +17%     │
  ├──────────────────────┼───────────┼──────────┼──────────┤
  │ 100B                 │ 482       │ 541      │ +12%     │
  ├──────────────────────┼───────────┼──────────┼──────────┤
  │ 1MB                  │ 6,511     │ 8,743    │ +34%     │
  ├──────────────────────┼───────────┼──────────┼──────────┤
  │ TSP Decrypt          │           │          │          │
  ├──────────────────────┼───────────┼──────────┼──────────┤
  │ 1B                   │ 492       │ 543      │ +10%     │
  ├──────────────────────┼───────────┼──────────┼──────────┤
  │ 100B                 │ 484       │ 533      │ +10%     │
  ├──────────────────────┼───────────┼──────────┼──────────┤
  │ 1MB                  │ 6,714     │ 8,659    │ +29%     │
  └──────────────────────┴───────────┴──────────┴──────────┘
  FFM vs TSC (Legacy Java SDK)
  ┌───────────────┬──────────┬──────────┬─────────────┐
  │   Operation   │ TSC (µs) │ FFM (µs) │ Improvement │
  ├───────────────┼──────────┼──────────┼─────────────┤
  │ encrypt 1B    │ 683      │ 549      │ -20%        │
  ├───────────────┼──────────┼──────────┼─────────────┤
  │ encrypt 100B  │ 739      │ 541      │ -27%        │
  ├───────────────┼──────────┼──────────┼─────────────┤
  │ encrypt 10KB  │ 2,372    │ 697      │ -71%        │
  ├───────────────┼──────────┼──────────┼─────────────┤
  │ encrypt 1MB   │ 13,105   │ 8,743    │ -33%        │
  ├───────────────┼──────────┼──────────┼─────────────┤
  │ decrypt 1B    │ 736      │ 543      │ -26%        │
  ├───────────────┼──────────┼──────────┼─────────────┤
  │ decrypt 100B  │ 741      │ 533      │ -28%        │
  ├───────────────┼──────────┼──────────┼─────────────┤
  │ decrypt 10KB  │ 2,328    │ 705      │ -70%        │
  ├───────────────┼──────────┼──────────┼─────────────┤
  │ decrypt 1MB   │ 13,149   │ 8,659    │ -34%        │
  ├───────────────┼──────────┼──────────┼─────────────┤
  │ batch 10×100B │ 3,549    │ 1,204    │ -66%        │
  └───────────────┴──────────┴──────────┴─────────────┘
  Rust Native Baselines
  ┌───────────────────────────┬───────────┐
  │         Operation         │ Time (µs) │
  ├───────────────────────────┼───────────┤
  │ Standalone Roundtrip      │           │
  ├───────────────────────────┼───────────┤
  │ 10B                       │ 7.6       │
  ├───────────────────────────┼───────────┤
  │ 10KB                      │ 99        │
  ├───────────────────────────┼───────────┤
  │ 100KB                     │ 930       │
  ├───────────────────────────┼───────────┤
  │ Standalone Vector Encrypt │           │
  ├───────────────────────────┼───────────┤
  │ d=384                     │ 12        │
  ├───────────────────────────┼───────────┤
  │ d=768                     │ 22        │
  ├───────────────────────────┼───────────┤
  │ d=1536                    │ 44        │
  ├───────────────────────────┼───────────┤
  │ d=2048                    │ 58        │
  ├───────────────────────────┼───────────┤
  │ TSP Encrypt               │           │
  ├───────────────────────────┼───────────┤
  │ 1B                        │ 471       │
  ├───────────────────────────┼───────────┤
  │ 100B                      │ 482       │
  ├───────────────────────────┼───────────┤
  │ 10KB                      │ 848       │
  ├───────────────────────────┼───────────┤
  │ 1MB                       │ 6,511     │
  └───────────────────────────┴───────────┘
```

For standalone the benefit is large and clear. For all the async stuff we have them pass in a foreign language native HTTP client, but clearly there's something causing additional slowness in the changes to the async implementation the AI came up with. There's strong enough indicators that it should be beneficial that I'm hoping to come back to it and do a real implementation.

On the Android problem, if we pursue this we'll likely conditionally generate https://github.com/vova7878/PanamaPort to try.

**Comment by skeet70** [comment]
I reimplemented Java FFM in a less proof-of-concept way, after updating the java bindings to 0.31.0, so I was able to use the benchmarks on main for some more concrete comparisons. This is all on a 2024 Macbook M4 Max w/ 36GB RAM:

### Function Calls (foreign code calling Rust)

Test Case | Java FFM | Java JNA | Kotlin (JNA) | Python | Swift
-- | -- | -- | -- | -- | --
call-only | 4.3 ns | 1.83 us | 2.89 us | 792 ns | 165 ns
primitives | 4.5 ns | 1.62 us | 3.27 us | 1.39 us | 191 ns
strings | 324 ns | 11.46 us | 15.08 us | 9.06 us | 929 ns
large-strings | 3.69 us | 15.63 us | 21.69 us | 12.33 us | 1.28 us
records | 371 ns | 11.62 us | 15.29 us | 18.66 us | 2.54 us
enums | 334 ns | 14.47 us | 15.41 us | 14.22 us | 2.04 us
vecs | 581 ns | 13.14 us | 15.36 us | 18.02 us | 4.55 us
hash-maps | 638 ns | 12.91 us | 15.48 us | 23.15 us | 6.39 us
interfaces | 530 ns * | 6.07 us | 105.60 us | 4.21 us | 409 ns
trait-interfaces | 532 ns * | 11.98 us | 150.70 us | 4.30 us | 477 ns
nested-data | 1.90 us | 12.56 us | 15.64 us | 77.09 us | 20.86 us
errors | 682 ns | 4.51 us | 6.44 us | 3.20 us | 646 ns

### Callbacks (Rust calling foreign code)

Test Case | Java FFM | Java JNA | Kotlin (JNA) | Python | Swift
-- | -- | -- | -- | -- | --
call-only | 55 ns | 1.98 us | 2.76 us | 492 ns | 121 ns
primitives | 59 ns | 2.66 us | 3.40 us | 804 ns | 166 ns
strings | 348 ns | 13.76 us | 15.69 us | 8.17 us | 811 ns
large-strings | 3.84 us | 17.47 us | 20.55 us | 10.77 us | 1.60 us
records | 449 ns | 11.55 us | 15.95 us | 14.26 us | 2.86 us
enums | 406 ns | 11.17 us | 14.93 us | 11.04 us | 2.47 us
vecs | 593 ns | 11.11 us | 15.02 us | 17.95 us | 5.01 us
hash-maps | 723 ns | 12.36 us | 15.72 us | 21.39 us | 7.16 us
interfaces | 964 ns * | 7.44 us | 408.65 us | 4.19 us | 448 ns
trait-interfaces | 790 ns * | 7.49 us | 361.69 us | 4.33 us | 552 ns
nested-data | 1.93 us | 19.77 us | 16.29 us | 67.20 us | 24.54 us
errors | 638 ns | 6.37 us | 7.98 us | 4.35 us | 574 ns


* Interface benchmarks have high variance due to (it seems like from how many objects are created) GC pauses (19-23% severe outliers); min values are ~250-470ns.

### Java FFM vs JNA Speedup

Test Case | Fn Call Speedup | Callback Speedup
-- | -- | --
call-only | 426x | 36x
primitives | 360x | 45x
strings | 35x | 40x
large-strings | 4.2x | 4.5x
records | 31x | 26x
enums | 43x | 28x
vecs | 23x | 19x
hash-maps | 20x | 17x
interfaces | 11x | 8x
trait-interfaces | 23x | 9x
nested-data | 6.6x | 10x
errors | 6.6x | 10x

**Comment by Sajjon** [comment]
@skeet70 

(Written on 1st of April, hope these numbers are real 😅)

> 426x

Wow that is no small feat, good job!

**Comment by bendk** [comment]
Really cool.  Did you keep the FFI more-or-less the same and just replaced the implementation?  I'm working through a JNI-based one that tries to make some fundamental changes.  I'm expecting a similar speedup, which means if you combined FFM and the new FFI those numbers would be even higher.

Also, how did you make those tables?  I'm hoping to make something similar myself.  I was experimenting with one way in #2845, but I didn't like the result so much.  Maybe we just need a regex to parse the criterion output and generate a nice table.

**Comment by Sajjon** [comment]
> Also, how did you make those tables?

Perhaps [Divan](https://github.com/nvzqz/divan) might be worth looking into, it has bench marking + table generations, I haven't used it personally though.

I've used [hotpath](https://hotpath.rs/) in the past, which I was quite happy with.

**Comment by skeet70** [comment]
> Really cool. Did you keep the FFI more-or-less the same and just replaced the implementation? I'm working through a JNI-based one that tries to make some fundamental changes. I'm expecting a similar speedup, which means if you combined FFM and the new FFI those numbers would be even higher.

Yeah, the Uniffi side and FFI is all untouched, the generated code implementation switched to FFM. The most bespoke part was generating a slabbed memory manager that's shared by the struct returns and call status', where it grabs an `ofAuto` chunk of GC memory and the doles that out until it's full and grabs another, to try to amortize GC costs over more calls. Theres a big comment on that class in [the PR](https://github.com/IronCoreLabs/uniffi-bindgen-java/pull/54/changes#diff-5118b1428bcc08332f6f95f9ea01ba511d59f4669bed5917b33380d39470ffb5R109) explaining all the other things I tried. Also [a cleaner that cleans a few references on `register` of something new](https://github.com/IronCoreLabs/uniffi-bindgen-java/pull/54/changes#diff-1708aaaf38977ef1e9b3eb1e98fd2d08836b9356fa4caa46555dec8df656a34bR37), because the background-thread-only cleaner wasn't keeping up during some of the benchmarks, leading to OOM. Everything else was a semi-straightforward switch.

> Also, how did you make those tables? I'm hoping to make something similar myself. I was experimenting with one way in [#2845](https://github.com/mozilla/uniffi-rs/pull/2845), but I didn't like the result so much. Maybe we just need a regex to parse the criterion output and generate a nice table.

Yeah, script w/regex to generate a nice table. Something like [this](https://gist.github.com/skeet70/5d1ceb9028fd389c424064e2c390c87d): `nix shell nixpkgs#python3 -c bench_compare.py "FFM" java_ffm_bench.txt "" /tmp/bench_python.txt "" /tmp/bench_kotlin.txt "" /tmp/bench_swift.txt` (cleaned up a bit and added grabbing the language from the results automatically, I was doing those manually initially)

**Comment by ShayBox** [comment]
I wasn't expecting FFM to be merged already! I'm not sure if this is the best place to discuss or not.

My project is in Kotlin, but I can use uniffi-bindgen-java if it means not dealing with cbindgen, jextract, and C ABI, but I had to learn the hard way bindgen doesn't work with cargo zigbuild, but I'm not going to investigate that further.

I build for 6 targets, but the uniffi dependency for macros seems to be depending on and causing uniffi-bindgen to build 6 times, even though I'm using uniffi-bindgen-java through Gradle.
